File: pocketr-java.api/src/main/java/com/decrux/pocketr/api/services/ledger/validations/TransactionAccountCurrencyValidator.java
package com.decrux.pocketr.api.services.ledger.validations;

import com.decrux.pocketr.api.entities.db.ledger.Account;
import com.decrux.pocketr.api.exceptions.BadRequestException;
import java.util.List;
import java.util.Objects;
import org.springframework.stereotype.Component;

@Component
public class TransactionAccountCurrencyValidator {

    public void validate(List<Account> accounts, String transactionCurrency) {
        for (Account account : accounts) {
            String accountCurrency = account.getCurrency() != null ? account.getCurrency().getCode() : null;
            if (!Objects.equals(accountCurrency, transactionCurrency)) {
                throw new BadRequestException(
                    "Account '"
                        + account.getName()
                        + "' has currency "
                        + accountCurrency
                        + " but transaction currency is "
                        + transactionCurrency
                );
            }
        }
    }
}

File: pocketr-api/src/main/kotlin/com/decrux/pocketr/api/services/ledger/validations/TransactionAccountCurrencyValidator.kt
package com.decrux.pocketr.api.services.ledger.validations

import com.decrux.pocketr.api.entities.db.ledger.Account
import com.decrux.pocketr.api.exceptions.BadRequestException
import org.springframework.stereotype.Component

@Component
class TransactionAccountCurrencyValidator {
    fun validate(
        accounts: List<Account>,
        transactionCurrency: String,
    ) {
        accounts.forEach { account ->
            if (account.currency?.code != transactionCurrency) {
                throw BadRequestException(
                    "Account '${account.name}' has currency ${account.currency?.code} but transaction currency is $transactionCurrency",
                )
            }
        }
    }
}
